src/streamlink/stream/dash.py

```diff
@@ -120,6 +122,13 @@ def iter_segments(self):
                         break
                     yield segment
 
+                    # init segments have duration set to None
+                    duration += segment.duration or 0.0
```

Comment:
Is there a chance that other segments have no duration set (after the init segment?). If so there could be a warning about it, because the overall duration is wrong in that case.

Reply:
The `duration` attribute is mandatory for segments in `SegmentList`s and `SegmentTemplate`s, and so is the `d` attribute in `SegmentTimeline`s.

See ISO/IEC 23009-1:2022 or the XML schema here:

- https://github.com/Dash-Industry-Forum/MPEG-Conformance-and-reference-source/blob/c50c32e61b5c1d3401840872a45bc10f6ea2709a/conformance/MPDValidator/schemas/DASH-MPD.xsd#L311
- https://github.com/Dash-Industry-Forum/MPEG-Conformance-and-reference-source/blob/c50c32e61b5c1d3401840872a45bc10f6ea2709a/conformance/MPDValidator/schemas/DASH-MPD.xsd#L371

Other kinds of DASH stream types are not supported by Streamlink.

We do however treat the `duration` attribute as optional:
- https://github.com/streamlink/streamlink/blame/6.0.1/src/streamlink/stream/dash_manifest.py#L733

This was added in 577617a2383cc75c6c94160b36aeccd7bd3b4bcf for some reason, so when I refactored the whole DASH code earlier this year, I kept it. While the specification needs to be strict, a DASH manifest parser shouldn't be, so even broken/invalid manifests can be played.

Each media `Segment` that gets yielded by the DASH manifest parser does have a duration set, and it can only be `None` if the `duration` attribute is missing, which is not Streamlink's fault in this case.

Reply:
Apparently the `duration` can indeed be optional, because of manifests with only a single segment.